File: python/problems/leetcode/0355_design_twitter/solution.py
"""LeetCode 355: Design Twitter."""

from __future__ import annotations

import heapq
from collections import defaultdict


class Twitter:
    def __init__(self) -> None:
        self._time = 0
        self._tweets: dict[int, list[tuple[int, int]]] = defaultdict(list)
        self._follows: dict[int, set[int]] = defaultdict(set)

    def postTweet(self, userId: int, tweetId: int) -> None:
        self._time += 1
        self._tweets[userId].append((self._time, tweetId))

    def getNewsFeed(self, userId: int) -> list[int]:
        users = self._follows[userId] | {userId}
        heap: list[tuple[int, int, int]] = []
        for u in users:
            if self._tweets[u]:
                t, tid = self._tweets[u][-1]
                heapq.heappush(heap, (-t, u, len(self._tweets[u]) - 1))
        res: list[int] = []
        while heap and len(res) < 10:
            negt, u, idx = heapq.heappop(heap)
            res.append(self._tweets[u][idx][1])
            if idx > 0:
                t2, _ = self._tweets[u][idx - 1]
                heapq.heappush(heap, (-t2, u, idx - 1))
        return res

    def follow(self, followerId: int, followeeId: int) -> None:
        if followerId != followeeId:
            self._follows[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        self._follows[followerId].discard(followeeId)


if __name__ == "__main__":
    tw = Twitter()
    tw.postTweet(1, 5)
    assert tw.getNewsFeed(1) == [5]
    tw.postTweet(1, 3)
    assert tw.getNewsFeed(1) == [3, 5]
    tw.follow(1, 2)
    tw.postTweet(2, 6)
    assert tw.getNewsFeed(1) == [6, 3, 5]
    tw.unfollow(1, 2)
    assert tw.getNewsFeed(1) == [3, 5]
    print("0355 OK")

File: cpp/problems/leetcode/0355_design_twitter/solution.cpp
// LeetCode 355: Design Twitter
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Twitter {
    int time = 0;
    unordered_map<int, vector<pair<int, int>>> tweets;
    unordered_map<int, unordered_set<int>> follows;

public:
    void postTweet(int userId, int tweetId) {
        tweets[userId].push_back({++time, tweetId});
    }

    vector<int> getNewsFeed(int userId) {
        unordered_set<int> users = follows[userId];
        users.insert(userId);
        priority_queue<tuple<int, int, int>> pq;
        for (int u : users) {
            if (!tweets[u].empty()) {
                int idx = (int)tweets[u].size() - 1;
                pq.push({tweets[u][idx].first, u, idx});
            }
        }
        vector<int> res;
        while (!pq.empty() && (int)res.size() < 10) {
            auto [t, u, idx] = pq.top();
            pq.pop();
            res.push_back(tweets[u][idx].second);
            if (idx > 0) pq.push({tweets[u][idx - 1].first, u, idx - 1});
        }
        return res;
    }

    void follow(int followerId, int followeeId) {
        if (followerId != followeeId) follows[followerId].insert(followeeId);
    }

    void unfollow(int followerId, int followeeId) {
        follows[followerId].erase(followeeId);
    }
};

int main() {
    Twitter tw;
    tw.postTweet(1, 5);
    assert(tw.getNewsFeed(1) == vector<int>({5}));
    tw.postTweet(1, 3);
    assert(tw.getNewsFeed(1) == vector<int>({3, 5}));
    tw.follow(1, 2);
    tw.postTweet(2, 6);
    assert(tw.getNewsFeed(1) == vector<int>({6, 3, 5}));
    tw.unfollow(1, 2);
    assert(tw.getNewsFeed(1) == vector<int>({3, 5}));
    cout << "0355 OK" << endl;
    return 0;
}
